File: Sources/AstrolabeProtocol/Core/RuntimeGeometry.swift
//
//  RuntimeGeometry.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/16.
//

public enum RuntimeMeasurementUnit: String, Codable, Sendable {
    case logical
    case scaledLogical
    case pixel
}

public enum RuntimeCoordinateSpace: String, Codable, Sendable {
    case local
    case parent
    case screen
    case viewport
}

public struct RuntimeMeasurement: Codable, Equatable, Sendable {
    /// Numeric magnitude in the declared unit.
    public let value: Double

    /// Unit used by the magnitude.
    public let unit: RuntimeMeasurementUnit

    public init(value: Double, unit: RuntimeMeasurementUnit) {
        self.value = value
        self.unit = unit
    }
}

public struct RuntimeCoordinatePoint: Codable, Equatable, Sendable {
    /// Horizontal coordinate.
    public let x: Double

    /// Vertical coordinate.
    public let y: Double

    /// Coordinate space containing the point.
    public let coordinateSpace: RuntimeCoordinateSpace

    /// Logical or physical unit used by the coordinates.
    public let unit: RuntimeMeasurementUnit

    public init(
        x: Double,
        y: Double,
        coordinateSpace: RuntimeCoordinateSpace,
        unit: RuntimeMeasurementUnit
    ) {
        self.x = x
        self.y = y
        self.coordinateSpace = coordinateSpace
        self.unit = unit
    }
}

public struct RuntimeMeasuredSize: Codable, Equatable, Sendable {
    /// Non-negative horizontal extent.
    public let width: Double

    /// Non-negative vertical extent.
    public let height: Double

    /// Logical or physical unit used by the extents.
    public let unit: RuntimeMeasurementUnit

    public init(width: Double, height: Double, unit: RuntimeMeasurementUnit) {
        self.width = width
        self.height = height
        self.unit = unit
    }
}

public struct RuntimeVector: Codable, Equatable, Sendable {
    /// Horizontal delta.
    public let dx: Double

    /// Vertical delta.
    public let dy: Double

    /// Logical or physical unit used by the deltas.
    public let unit: RuntimeMeasurementUnit

    public init(dx: Double, dy: Double, unit: RuntimeMeasurementUnit) {
        self.dx = dx
        self.dy = dy
        self.unit = unit
    }
}

public struct RuntimeInsets: Codable, Equatable, Sendable {
    /// Top inset.
    public let top: Double

    /// Left inset.
    public let left: Double

    /// Bottom inset.
    public let bottom: Double

    /// Right inset.
    public let right: Double

    /// Logical or physical unit used by every edge.
    public let unit: RuntimeMeasurementUnit

    public init(
        top: Double,
        left: Double,
        bottom: Double,
        right: Double,
        unit: RuntimeMeasurementUnit
    ) {
        self.top = top
        self.left = left
        self.bottom = bottom
        self.right = right
        self.unit = unit
    }
}

public struct RuntimeCoordinateRect: Codable, Equatable, Sendable {
    /// Horizontal origin.
    public let x: Double

    /// Vertical origin.
    public let y: Double

    /// Non-negative horizontal extent.
    public let width: Double

    /// Non-negative vertical extent.
    public let height: Double

    /// Coordinate space containing the rectangle.
    public let coordinateSpace: RuntimeCoordinateSpace

    /// Logical or physical unit used by the rectangle.
    public let unit: RuntimeMeasurementUnit

    public init(
        x: Double,
        y: Double,
        width: Double,
        height: Double,
        coordinateSpace: RuntimeCoordinateSpace,
        unit: RuntimeMeasurementUnit
    ) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.coordinateSpace = coordinateSpace
        self.unit = unit
    }
}

public struct RuntimeScale: Codable, Equatable, Sendable {
    /// Horizontal logical-to-pixel scale.
    public let x: Double

    /// Vertical logical-to-pixel scale.
    public let y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

public struct RuntimeDisplayInfo: Codable, Equatable, Sendable {
    /// Display dimensions in logical layout units.
    public let logicalSize: RuntimeMeasuredSize

    /// Display dimensions in physical pixels.
    public let pixelSize: RuntimeMeasuredSize

    /// Per-axis conversion from logical units to pixels.
    public let logicalToPixelScale: RuntimeScale

    /// Maximum refresh rate when reported by the platform.
    @RuntimeRequiredNullable public private(set) var maximumRefreshRate: Double?

    public init(
        logicalSize: RuntimeMeasuredSize,
        pixelSize: RuntimeMeasuredSize,
        logicalToPixelScale: RuntimeScale,
        maximumRefreshRate: Double?
    ) {
        self.logicalSize = logicalSize
        self.pixelSize = pixelSize
        self.logicalToPixelScale = logicalToPixelScale
        self.maximumRefreshRate = maximumRefreshRate
    }
}

File: Implementations/Kotlin/src/main/kotlin/dev/astrolabe/protocol/core/RuntimeGeometry.kt
//
//  RuntimeGeometry.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/20.
//

package dev.astrolabe.protocol

import kotlinx.serialization.Serializable

/** Unit used by runtime measurements. */
@Serializable
public enum class RuntimeMeasurementUnit {
    logical,
    scaledLogical,
    pixel
}

/** Coordinate space containing a runtime geometry value. */
@Serializable
public enum class RuntimeCoordinateSpace {
    local,
    parent,
    screen,
    viewport
}

/** Scalar measurement with an explicit unit. */
@Serializable
public data class RuntimeMeasurement(
    /** Numeric magnitude. */
    public val value: Double,
    /** Unit used by the magnitude. */
    public val unit: RuntimeMeasurementUnit
)

/** Point in an explicitly declared coordinate space. */
@Serializable
public data class RuntimeCoordinatePoint(
    /** Horizontal coordinate. */
    public val x: Double,
    /** Vertical coordinate. */
    public val y: Double,
    /** Coordinate space containing the point. */
    public val coordinateSpace: RuntimeCoordinateSpace,
    /** Unit used by both coordinates. */
    public val unit: RuntimeMeasurementUnit
) {
    init {
        require(unit != RuntimeMeasurementUnit.scaledLogical) {
            "Coordinate point unit must be logical or pixel"
        }
    }
}

/** Two-dimensional measured extent. */
@Serializable
public data class RuntimeMeasuredSize(
    /** Non-negative horizontal extent. */
    public val width: Double,
    /** Non-negative vertical extent. */
    public val height: Double,
    /** Unit used by both extents. */
    public val unit: RuntimeMeasurementUnit
) {
    init {
        require(width >= 0.0 && height >= 0.0) { "Measured size dimensions cannot be negative" }
        require(unit != RuntimeMeasurementUnit.scaledLogical) {
            "Measured size unit must be logical or pixel"
        }
    }
}

/** Two-dimensional signed delta. */
@Serializable
public data class RuntimeVector(
    /** Horizontal delta. */
    public val dx: Double,
    /** Vertical delta. */
    public val dy: Double,
    /** Unit used by both deltas. */
    public val unit: RuntimeMeasurementUnit
) {
    init {
        require(unit != RuntimeMeasurementUnit.scaledLogical) {
            "Vector unit must be logical or pixel"
        }
    }
}

/** Insets measured from four edges. */
@Serializable
public data class RuntimeInsets(
    /** Top inset. */
    public val top: Double,
    /** Left inset. */
    public val left: Double,
    /** Bottom inset. */
    public val bottom: Double,
    /** Right inset. */
    public val right: Double,
    /** Unit used by every edge. */
    public val unit: RuntimeMeasurementUnit
) {
    init {
        require(unit != RuntimeMeasurementUnit.scaledLogical) {
            "Insets unit must be logical or pixel"
        }
    }
}

/** Rectangle in an explicitly declared coordinate space. */
@Serializable
public data class RuntimeCoordinateRect(
    /** Horizontal origin. */
    public val x: Double,
    /** Vertical origin. */
    public val y: Double,
    /** Non-negative horizontal extent. */
    public val width: Double,
    /** Non-negative vertical extent. */
    public val height: Double,
    /** Coordinate space containing the rectangle. */
    public val coordinateSpace: RuntimeCoordinateSpace,
    /** Unit used by all rectangle components. */
    public val unit: RuntimeMeasurementUnit
) {
    init {
        require(width >= 0.0 && height >= 0.0) { "Rectangle dimensions cannot be negative" }
        require(unit != RuntimeMeasurementUnit.scaledLogical) {
            "Rectangle unit must be logical or pixel"
        }
    }
}

/** Per-axis logical-to-pixel conversion. */
@Serializable
public data class RuntimeScale(
    /** Horizontal scale. */
    public val x: Double,
    /** Vertical scale. */
    public val y: Double
) {
    init {
        require(x > 0.0 && y > 0.0) { "Scale components must be greater than zero" }
    }
}

/** Display facts required for coordinate conversion. */
@Serializable
public data class RuntimeDisplayInfo(
    /** Display dimensions in logical layout units. */
    public val logicalSize: RuntimeMeasuredSize,
    /** Display dimensions in physical pixels. */
    public val pixelSize: RuntimeMeasuredSize,
    /** Per-axis conversion from logical units to pixels. */
    public val logicalToPixelScale: RuntimeScale,
    /** Maximum refresh rate when reported by the platform. */
    public val maximumRefreshRate: Double?
) {
    init {
        require(logicalSize.unit == RuntimeMeasurementUnit.logical) {
            "Logical display size must use logical units"
        }
        require(pixelSize.unit == RuntimeMeasurementUnit.pixel) {
            "Pixel display size must use pixel units"
        }
        require(maximumRefreshRate == null || maximumRefreshRate > 0.0) {
            "Maximum refresh rate must be greater than zero"
        }
    }
}
